File: leetcode/0001-1000/0004-median-of-two-sorted-arrays/MedianOfTwoSortedArrays.ts
function findMedianSortedArrays(nums1: number[], nums2: number[]): number {
  let current = 0;
  let previous = 0;

  const n = (nums1.length + nums2.length + 1) / 2;
  let i = 0;
  let j = 0;

  for (let index = 0; index < n; index++) {
    previous = current;

    if (i >= nums1.length) {
      current = nums2[j++];
    } else if (j >= nums2.length) {
      current = nums1[i++];
    } else {
      current = nums1[i] >= nums2[j] ? nums2[j++] : nums1[i++];
    }
  }

  if ((nums1.length + nums2.length) % 2 === 1) {
    previous = current;
  }

  return (current + previous) / 2;
};

// 0, 1, 1, 2, 2, 3, 3, 4
console.log(findMedianSortedArrays([0, 1, 2, 3], [1, 2, 3, 4]))
// -1, 0, 1, 1, 2, 2, 3, 3, 4
console.log(findMedianSortedArrays([-1, 0, 1, 2, 3], [1, 2, 3, 4]))
// -2, -1, 0, 1, 1, 2, 2, 3, 3, 4
console.log(findMedianSortedArrays([-2, -1, 0, 1, 2, 3], [1, 2, 3, 4]))

File: leetcode/0001-1000/0001-0100/0001-0010/0004-median-of-two-sorted-arrays/MedianOfTwoSortedArrays.java
public class MedianOfTwoSortedArrays {

  public static void main(String[] args) {
    var solution = new Solution();

    System.out.println(solution.findMedianSortedArrays(
        new int[] { 1, 2 }, new int[] { 3, 4 }));

    System.out.println(solution.findMedianSortedArrays(
        new int[] { 1, 3 }, new int[] {}));
  }
}

class Solution {
  public double findMedianSortedArrays(int[] nums1, int[] nums2) {
    var m = nums1.length;
    var n = nums2.length;

    var current = 0;
    var previous = 0;
    var first = 0;
    var second = 0;

    for (var i = 0; i <= (m + n) / 2; i++) {
      if (first >= nums1.length) {
        previous = current;
        current = nums2[second];

        second++;
      } else if (second >= nums2.length) {

        previous = current;
        current = nums1[first];

        first++;
      } else if (nums1[first] > nums2[second]) {
        previous = current;
        current = nums2[second];
        second++;
      } else {
        previous = current;
        current = nums1[first];
        first++;
      }
    }

    if ((m + n) % 2 == 1) {
      previous = current;
    }

    return 0.5 * (previous + current);
  }
}
